File: frontend/lib/ai-extraction-throttle.ts
// Shared per-user rate-limit gate for the AI extraction BFF routes
// (extract-show + extract-collection, PSY-855). Both routes call enforceThrottle
// BEFORE doing any (paid) Anthropic work; on a limit hit it returns the 429
// envelope the routes must send back WITHOUT calling Anthropic.
//
// The counter is Postgres-backed in the Go backend — not in-memory — so the
// limit holds across Vercel serverless instances (an in-memory counter would
// let a user multiply the limit by the number of warm instances). This helper
// just forwards the user's auth_token cookie to the backend throttle endpoint;
// the backend identifies the user from the JWT (no spoofable user_id), counts
// the attempt, and bypasses admins (PSY-345).

import { NextResponse } from 'next/server'
import * as Sentry from '@sentry/nextjs'

const BACKEND_URL = process.env.BACKEND_URL || 'http://localhost:8080'

interface ThrottleDecision {
  allowed: boolean
  retry_after_seconds: number
  limit: number
  window_seconds: number
}

// Shape the user-facing 429 body. Mirrors PSY-589's backend Huma 429 pattern:
// a human-readable `error` string plus a machine-readable `retry_after` (seconds).
interface ThrottleErrorBody {
  success: false
  error: string
  retry_after: number
}

/**
 * Build the "try again in N minutes" copy from the seconds-remaining. Rounds UP
 * to the next whole minute so the hint never undershoots the real cooldown; for
 * sub-minute waits it reads in seconds so a 30s wait doesn't round to a
 * misleading "1 minute".
 */
function formatRetryHint(retryAfterSeconds: number): string {
  const seconds = Math.max(1, Math.ceil(retryAfterSeconds))
  if (seconds < 60) {
    const unit = seconds === 1 ? 'second' : 'seconds'
    return `Rate limit exceeded. Try again in ${seconds} ${unit}.`
  }
  const minutes = Math.ceil(seconds / 60)
  const unit = minutes === 1 ? 'minute' : 'minutes'
  return `Rate limit exceeded. Try again in ${minutes} ${unit}.`
}

/**
 * Enforce the per-user AI-extraction rate limit. Returns `{ ok: true }` when the
 * attempt is allowed (the backend has already counted it), or `{ ok: false,
 * response }` carrying the exact response the route must return instead of
 * calling Anthropic:
 *   - 429 (limit hit): Retry-After header + the decided JSON body.
 *   - 503 (gate unavailable): fail CLOSED — if the counter is down we must not
 *     hand out a free pass that would drain the Anthropic budget.
 *
 * `authToken` is the caller's auth_token cookie value (already extracted by the
 * route's auth gate). The backend throttle endpoint is on the Protected group,
 * so the cookie identifies the user.
 */
export async function enforceThrottle(
  authToken: string,
  sentryService: string
): Promise<{ ok: true } | { ok: false; response: NextResponse }> {
  let res: Response
  try {
    res = await fetch(`${BACKEND_URL}/ai-extraction/throttle`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Cookie: `auth_token=${authToken}`,
      },
    })
  } catch (error) {
    // Backend unreachable — fail closed. A network blip on the gate must not
    // become an open door to the paid Anthropic API.
    Sentry.captureException(error, {
      level: 'error',
      tags: { service: sentryService, error_type: 'throttle_unreachable' },
    })
    return {
      ok: false,
      response: NextResponse.json(
        { success: false, error: 'Rate limit check temporarily unavailable. Please try again.' },
        { status: 503 }
      ),
    }
  }

  if (!res.ok) {
    // Backend returned a non-2xx (e.g. 503 from the handler when the counter
    // write failed, or 401 if the cookie was rejected). Fail closed: don't call
    // Anthropic. Pass the status through so a 401 stays a 401.
    Sentry.captureMessage(`AI-extraction throttle gate returned ${res.status}`, {
      level: res.status >= 500 ? 'error' : 'warning',
      tags: { service: sentryService, error_type: 'throttle_gate_error' },
    })
    const status = res.status === 401 ? 401 : 503
    const message =
      status === 401
        ? 'Authentication required'
        : 'Rate limit check temporarily unavailable. Please try again.'
    return {
      ok: false,
      response: NextResponse.json({ success: false, error: message }, { status }),
    }
  }

  const decision = (await res.json().catch(() => null)) as ThrottleDecision | null
  if (!decision || typeof decision.allowed !== 'boolean') {
    // Malformed decision — fail closed.
    return {
      ok: false,
      response: NextResponse.json(
        { success: false, error: 'Rate limit check temporarily unavailable. Please try again.' },
        { status: 503 }
      ),
    }
  }

  if (decision.allowed) {
    return { ok: true }
  }

  const retryAfter = Math.max(1, Math.ceil(decision.retry_after_seconds || 1))
  const body: ThrottleErrorBody = {
    success: false,
    error: formatRetryHint(retryAfter),
    retry_after: retryAfter,
  }
  return {
    ok: false,
    response: NextResponse.json(body, {
      status: 429,
      headers: { 'Retry-After': String(retryAfter) },
    }),
  }
}

File: backend/internal/services/ratelimit/ai_extraction_throttle.go
// Package ratelimit holds the Postgres-backed per-user rate limiter for the AI
// extraction routes (extract-show + extract-collection, PSY-855).
//
// The two Next.js BFF extract routes proxy to this backend before doing any
// (paid) Anthropic work. The counter MUST be Postgres-backed — not in-memory —
// because the routes run on Vercel serverless instances that don't share
// process memory, so an in-memory counter would let a user multiply the limit
// by the number of warm instances.
package ratelimit

import (
	"fmt"
	"math"
	"time"

	"gorm.io/gorm"

	"psychic-homily-backend/db"
)

const (
	// AIExtractionLimit is the number of AI extractions a single non-admin user
	// may run per rolling window. Applies identically to both extract routes —
	// they share one counter so a user can't get 10 of each (PSY-855 spec'd a
	// single 10/hr budget).
	AIExtractionLimit = 10

	// AIExtractionWindow is the rolling window the limit applies over.
	AIExtractionWindow = time.Hour
)

// ThrottleDecision is the result of an increment-and-check. Allowed is false
// once the user has exceeded AIExtractionLimit within the current window;
// RetryAfterSeconds is then the whole-second wait until the window resets
// (always >= 1 when Allowed is false).
type ThrottleDecision struct {
	Allowed           bool
	RetryAfterSeconds int
}

// AIExtractionThrottleService gates the AI extraction routes behind a per-user
// rolling-window counter.
type AIExtractionThrottleService struct {
	db *gorm.DB
}

// NewAIExtractionThrottleService creates the throttle service.
func NewAIExtractionThrottleService(database *gorm.DB) *AIExtractionThrottleService {
	if database == nil {
		database = db.GetDB()
	}
	return &AIExtractionThrottleService{db: database}
}

// scanRow is the post-write window state returned by the atomic upsert.
type scanRow struct {
	WindowStart  time.Time `gorm:"column:window_start"`
	RequestCount int       `gorm:"column:request_count"`
}

// CheckAndIncrement atomically records one extraction attempt for the user and
// reports whether it is allowed under the limit.
//
// Atomicity matters: two concurrent serverless instances must not both read
// count=9, both decide "allowed", and both write count=10 — that would let 11
// through. The whole reset-or-increment decision happens in a SINGLE upsert
// statement (the CASE in the ON CONFLICT clause), so Postgres row locking
// serializes concurrent callers and RETURNING hands back the authoritative
// post-write state in the same round-trip.
//
// The window resets in-place when it has elapsed: if NOW() >= window_start +
// window, the row's window_start becomes NOW() and request_count becomes 1.
// Otherwise request_count is incremented. The caller is allowed iff the
// resulting count is within AIExtractionLimit.
//
// Note: a blocked attempt still increments the count. That's harmless —
// RetryAfterSeconds is derived from window_start (not the count), so a user
// hammering the route past the limit doesn't extend their own cooldown.
func (s *AIExtractionThrottleService) CheckAndIncrement(userID uint) (ThrottleDecision, error) {
	if s.db == nil {
		return ThrottleDecision{}, fmt.Errorf("database not initialized")
	}
	if userID == 0 {
		return ThrottleDecision{}, fmt.Errorf("userID is required")
	}

	windowSeconds := int(AIExtractionWindow / time.Second)

	var row scanRow
	// The interval is built from windowSeconds (a constant int) via make_interval,
	// not string-interpolated, so there's no injection surface; userID is bound.
	err := s.db.Raw(`
		INSERT INTO ai_extraction_throttle (user_id, window_start, request_count, created_at, updated_at)
		VALUES (?, NOW(), 1, NOW(), NOW())
		ON CONFLICT (user_id) DO UPDATE SET
			window_start = CASE
				WHEN ai_extraction_throttle.window_start + make_interval(secs => ?) <= NOW()
					THEN NOW()
				ELSE ai_extraction_throttle.window_start
			END,
			request_count = CASE
				WHEN ai_extraction_throttle.window_start + make_interval(secs => ?) <= NOW()
					THEN 1
				ELSE ai_extraction_throttle.request_count + 1
			END,
			updated_at = NOW()
		RETURNING window_start, request_count
	`, userID, windowSeconds, windowSeconds).Scan(&row).Error
	if err != nil {
		return ThrottleDecision{}, fmt.Errorf("failed to record extraction attempt: %w", err)
	}

	if row.RequestCount <= AIExtractionLimit {
		return ThrottleDecision{Allowed: true}, nil
	}

	// Over the limit — compute the whole-second wait until the window resets.
	// Clamp to >= 1 so the Retry-After header / "try again in N" copy never
	// reads "0 seconds" due to sub-second rounding at the window boundary.
	resetAt := row.WindowStart.Add(AIExtractionWindow)
	retryAfter := int(math.Ceil(time.Until(resetAt).Seconds()))
	if retryAfter < 1 {
		retryAfter = 1
	}

	return ThrottleDecision{Allowed: false, RetryAfterSeconds: retryAfter}, nil
}
